src/FreeformIdeas.js

```diff
@@ -1,90 +1,173 @@
 
-  const [selectedIdea, setSelectedIdea] = useState(ideas.a);
-  const setSelected = useCallback((id) => setSelectedIdea(ideas[id]), [ideas]);
+  const [selectedId, setSelectedId] = useState("a");
+  const selectedIdea = ideas[selectedId];
+
+  const [linkerEnd, setLinkerEnd] = useState();
+  const [linkingState, setLinkingState] = useState(false);
+  const linkBreak = useCallback(() => setLinkingState(false), []);
+
+  const onLinkStart = useCallback((e) => {
+    setLinkerEnd([e.clientY - 65, e.clientX]);
+    setLinkingState(true);
+  }, []);
+
+  const adjustLinker = useCallback(
+    (e) => {
+      if (linkingState) {
+        setLinkerEnd([e.clientY - 65, e.clientX]);
+      }
+    },
+    [linkingState]
+  );
+
+  const onLinkEnd = useCallback(
+    (id) => {
+      if (selectedIdea.id !== id && linkingState) {
+        if (!selectedIdea.linkList.includes(id)) {
+          selectedIdea.linkList.push(id);
+        }
+      }
+      setLinkingState(false);
+    },
+    [selectedIdea, linkingState]
+  );
 
   const [, drop] = useDrop({
     accept: ItemTypes.IDEA,
     drop(item, monitor) {
       const delta = monitor.getDifferenceFromInitialOffset();
       const left = Math.round(item.left + delta.x);
       const top = Math.round(item.top + delta.y);
-      setIdeas(
-        update(ideas, {
-          [item.id]: {
-            $merge: { left, top },
-          },
-        })
-      );
+      ideasDispatch({ type: "update", id: item.id, data: { left, top } });
     },
   });
 
   const editIdea = useCallback(
     (id) => {
-      setSelectedIdea(ideas[id]);
+      setSelectedId(id);
       ideaModalShow(true);
     },
-    [ideas, ideaModalShow]
+    [setSelectedId, ideaModalShow]
   );
 
   const handleDoubleClick = useCallback(
     (e) => {
       if (e.target.className !== "FreeformMap") return;
       e.preventDefault();
-      setSelectedIdea({
+      setCoords({
         top: e.nativeEvent.layerY,
         left: e.nativeEvent.layerX,
-        title: "",
       });
+      setSelectedId(uuidv4());
       setTimeout(ideaModalShow, 300);
     },
     [ideaModalShow]
   );
 
   const handleIdeaChange = useCallback(
     (idea) => {
+      if (idea.title === "") return;
+      setShowIdeaModal(false);
       if (!idea.id) {
-        idea = update(idea, { id: { $set: uuidv4() } });
-        setSelectedIdea(idea);
+        ideasDispatch({
+          type: "create",
+          id: selectedId,
+          data: update(idea, { $merge: coords }),
+        });
+        return;
       }
-      const { id, title } = idea;
-      if (title === "") return;
-      setIdeas(
-        update(ideas, {
-          [id]: {
-            $set: idea,
-          },
-        })
-      );
-      setShowIdeaModal(false);
+      ideasDispatch({ type: "update", id: selectedId, data: idea });
     },
-    [ideas]
+    [coords, selectedId]
   );
 
+  function calcCoords(height, width, arrowRotation, ideaTop, ideaLeft) {
```

Comment:
This function looks like it doesn't reference anything outside of its arguments. I would move this out of the component (cut and paste it to line 10) since it doesn't need to be here.

Reply:
Should the coords be rounded?

Reply:
Lastly, since this is a straight-up function, this is a great opportunity for us to write tests. Let's pair on adding tests for this.

Reply:
So for clarification. When a function can stand on its own other than what is given to it, it should be taken out of the component (in this case `FreeFormIdeas`)? While I think i understand why it works, is there a reason for this? Is it for a case where we have more than one component or something?

is there a benefit to rounding? with the inability to manage partial pixels, I figured it would just round on its own?

Are these tests unit tests? Ive seen people talk about unit tests, but dont really know what it is

Reply:
The more we add to the component, the harder it is to read we should only include what we need and break things out when it gets too big.

On top of that, we're technically redefining the function every time the component renders. If the function lives outside of the component, then it's only defined once when the file is loaded.

Lastly, you're right that we could use this function in other components (though I'd move it to a shared file at that point) but the main benefit is being able to easily test it outside of the context of the component.

Reply:
If it's working, it's probably rounding on its own, but since we want whole numbers by the end anyways, it'll make it easier to test if we have them, now.

And yes, the tests will be unit tests.